### PR TITLE
feat: add support for the `ceph-mds` relation

### PR DESCRIPTION
**Pull request by jedel1043**
# Description

This change allows integrating with `microceph` using the `ceph-mds` relation, which is required by charms such as `ceph-fs` to create filesystems on the cluster.
Do note that this only allows creating filesystems; a separate PR (#71) is in place to allow creating cephfs clients using broker requests.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] CleanCode (Code refactor, test updates, does not introduce functional changes)
- [ ] Documentation update (Doc only change)

## How Has This Been Tested?

> **_NOTE:_** All functional changes should accompany corresponding tests (unit tests, functional tests etc).

A functional test on the CI pipeline.

## Contributor's Checklist

Please check that you have:

- [x] self-reviewed the code in this PR.
- [x] added code comments, particularly in hard-to-understand areas.
- [x] updated the user documentation with corresponding changes.
- [x] added tests to verify effectiveness of this change.

### CODE REVIEW

File: src/relation_handlers.py [hunk 1]
```diff
@@ -669,3 +669,45 @@ def update_broker_data(self, data, event):
...
+
+    @staticmethod
+    def _select_relation(relations, relation_id):
+        for relation in relations:
+            if relation.id == relation_id:
+                return relation
+
+    @staticmethod
+    def _remote_unit_name(client_name):
+        return "ceph-radosgw/" + client_name.split("-")[-1]
```
Comment (lmlg) [inline src/relation_handlers.py:689]: wrong unit name (should be `ceph-fs`)

Comment (jedel1043) [reply src/relation_handlers.py:689]: Will remove. I think it's dead code anyways.

**Comment by jedel1043** [comment]
Not sure about how to go over testing this. I have local deployments using `ceph-fs`, but I think mocking it using `ops` would require adding a new relation library to `charm-ops-inteface-ceph-client`, which probaby needs its own design.

**Comment by lmlg** [comment]
> Not sure about how to go over testing this. I have local deployments using `ceph-fs`, but I think mocking it using `ops` would require adding a new relation library to `charm-ops-inteface-ceph-client` which probaby needs its own design.

A functional test would do the job. Essentially, we would need a model to be deployed where `ceph-fs` is related to `microceph` and one simple command that requires authorization completes successfully.

**Comment by jedel1043** [comment]
Added a functional test for the `mds` relation.

**Comment by UtkarshBhatthere** [comment]
/canonical/self-hosted-runners/run-workflows 6b54e9148de2fac91d25cd3a8e29d5d630012792

**Comment by UtkarshBhatthere** [comment]
@jedel1043 I have merged some CI patches, rebasing with main should make the tests better.

**Comment by jedel1043** [comment]
Rebased on main. I think it needs another workflow approval.

**Comment by UtkarshBhatthere** [comment]
/canonical/self-hosted-runners/run-workflows adfada1ca23292f2fbd7bd5f4b3b77a484b95b45

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -416,4 +416,61 @@ jobs:
...
+          # wait for charm to bootstrap.
+          juju wait-for unit ceph-fs/0 --query='workload-status=="blocked"' --timeout=30m
+          juju integrate ceph-fs:ceph-mds microceph:mds
+          # wait for filesystem creation
+          juju wait-for application ceph-fs --query='name=="ceph-fs" && (status=="active" || status=="idle")' --timeout=10m
+          filesystems=$(juju ssh microceph/0 -- "sudo microceph.ceph fs ls")
+          if ! echo "$filesystems" | grep -E "ceph-fs" ; then
+            echo "Failed to create the ceph-fs filesystem"
+            juju ssh microceph/0 -- "sudo microceph.ceph fs ls"
+            exit 1
```
Comment (jedel1043) [inline .github/workflows/build-and-test.yml:474]: I missed the `fi` after this line :(

**Comment by UtkarshBhatthere** [comment]
@jedel1043 from a cursory look, appears to be a trivial error in mds test. Please check.

**Comment by UtkarshBhatthere** [comment]
/canonical/self-hosted-runners/run-workflows 9349bc4c6d99fb8025c9f4560376d27d1f69ad5d

**Comment by sabaini** [comment]
/canonical/self-hosted-runners/run-workflows 9349bc4

**Comment by jedel1043** [comment]
No idea why the `Juju upgrade test` job is failing. Reading the logs, it seems like a unit is not upgrading the Microceph snap to `reef/stable` correctly? Though, that problem seems to be external to this PR.

**Comment by UtkarshBhatthere** [comment]
/canonical/self-hosted-runners/run-workflows e8ed07542974a8adb3128126488a1a9724cf6769

File: src/ceph_broker.py [hunk 3]
```diff
@@ -457,8 +460,7 @@ def get_named_key(name, caps=None, pool_list=None):
     :param caps: dict of cephx capabilities
     :returns: Returns a cephx key
     """
-    key_name = "client.{}".format(name)
-    key = ceph_auth_get(key_name)
+    key = ceph_auth_get(name)
     if key:
```
Comment (UtkarshBhatthere) [inline src/ceph_broker.py:464]: Hmm, this Key format change may cause breakage for existing use-cases ?

**Review by lmlg — APPROVED** [review]
LGTM, great stuff :)

**Pull request merged**